From 0df5618bfc8fedca889a0436e57515b52bd26c3f Mon Sep 17 00:00:00 2001
From: Pan Junzhong <junzhong.pan@spacemit.com>
Date: Wed, 8 May 2024 18:34:48 +0800
Subject: [PATCH 0643/1205] usb: k1x_udc_core: remove req from queue even it's
 already unmapped

done() is always called with req from queue, so it's safe to delete
it before checking if it's already unmapped. Add WARN_ON to gather
information when error occurs.

Change-Id: I20a03d4fb802d6ceed257164e0d87efb1b17953a
---
 drivers/usb/gadget/udc/k1x_udc_core.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/drivers/usb/gadget/udc/k1x_udc_core.c b/drivers/usb/gadget/udc/k1x_udc_core.c
index 0903113c10fd..7362f4ea6ca6 100755
--- a/drivers/usb/gadget/udc/k1x_udc_core.c
+++ b/drivers/usb/gadget/udc/k1x_udc_core.c
@@ -266,15 +266,16 @@ static int done(struct mv_ep *ep, struct mv_req *req, int status)
 
 	udc = (struct mv_udc *)ep->udc;
 
+	/* Removed the req from mv_ep->queue */
+	list_del_init(&req->queue);
+
 	if (req->req.dma == DMA_ADDR_INVALID && req->mapped == 0) {
 		dev_info(&udc->dev->dev, "%s request %p already unmapped",
 					ep->name, req);
+		WARN_ON_ONCE(1);
 		return -ESHUTDOWN;
 	}
 
-	/* Removed the req from fsl_ep->queue */
-	list_del_init(&req->queue);
-
 	/* req.status should be set as -EINPROGRESS in ep_queue() */
 	if (req->req.status == -EINPROGRESS)
 		req->req.status = status;
-- 
2.47.0

